lib/src/piece/piece.dart

```diff
@@ -210,7 +208,26 @@ class TextPiece extends Piece {
   }
 
   @override
-  String toString() => '`${_lines.join('¬')}`${_containsNewline ? '!' : ''}';
+  String toString() => '`${_lines.join('¬')}`';
+}
+
+/// A single line of text within a [TextPiece].
+class _Line {
+  String _text = '';
+
+  /// Whether this line should start at column one or use the surrounding
+  /// indentation.
+  final bool _isFlushLeft;
+
+  _Line({required bool flushLeft}) : _isFlushLeft = flushLeft;
+
+  void append(String text) {
+    // TODO(perf): Consider a faster way of accumulating text.
```

Comment:
Adding the `StringBuffer` now should be trivial - I think the introduction of a new abstraction for a line is a good reason to handle this TODO.

Reply:
Trivial, yes, but I don't know whether it will help or harm performance. Every single lexeme in the source program will go through this function, so it's a performance bottleneck. 99+% of TextPieces will only ever have a single token added so the overhead of creating a StringBuffer might be net loss.

I left this comment in here so that I can circle back and investigate it once I have enough of the language supported to run real benchmarks.